File: android/app/src/main/kotlin/com/thecityandthebike/ui/components/CalendarEntry.kt
package com.thecityandthebike.ui.components

import android.net.Uri
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage

data class CalendarPhoto(
    val submissionId: String,
    val imageUri: Uri
)

@Composable
fun CalendarEntry(
    dateLabel: String,
    yearLabel: String?,
    photos: List<CalendarPhoto>,
    onImageClick: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp)
    ) {
        Column(
            modifier = Modifier
                .width(56.dp)
                .padding(top = 4.dp)
        ) {
            Text(
                text = dateLabel,
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            if (yearLabel != null) {
                Text(
                    text = yearLabel,
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp)
        ) {
            photos.chunked(3).forEach { rowPhotos ->
                Row(
                    horizontalArrangement = Arrangement.spacedBy(2.dp)
                ) {
                    rowPhotos.forEach { photo ->
                        Box(
                            modifier = Modifier
                                .weight(1f)
                                .aspectRatio(1f)
                                .background(MaterialTheme.colorScheme.surfaceVariant)
                                .clickable(onClickLabel = "View photo") {
                                    onImageClick(photo.submissionId)
                                }
                        ) {
                            AsyncImage(
                                model = photo.imageUri,
                                contentDescription = "Captured image",
                                modifier = Modifier.matchParentSize(),
                                contentScale = ContentScale.Crop
                            )
                        }
                    }
                    // Fill remaining space in incomplete rows
                    repeat(3 - rowPhotos.size) {
                        Box(modifier = Modifier.weight(1f))
                    }
                }
            }
        }
    }
}

File: ios/Packages/TCATBFeed/Sources/TCATBFeed/Components/CalendarEntry.swift
import SwiftUI

public struct CalendarPhoto: Identifiable {
    public let id: String
    public let submissionId: String
    public let imageURL: URL

    public init(submissionId: String, imageURL: URL) {
        self.id = submissionId
        self.submissionId = submissionId
        self.imageURL = imageURL
    }
}

public struct CalendarEntry: View {
    let dateLabel: String
    let yearLabel: String?
    let photos: [CalendarPhoto]
    var onImageClick: (String) -> Void

    public init(
        dateLabel: String,
        yearLabel: String? = nil,
        photos: [CalendarPhoto],
        onImageClick: @escaping (String) -> Void
    ) {
        self.dateLabel = dateLabel
        self.yearLabel = yearLabel
        self.photos = photos
        self.onImageClick = onImageClick
    }

    public var body: some View {
        HStack(alignment: .top) {
            VStack(alignment: .leading) {
                Text(dateLabel)
                    .font(.caption)
                    .foregroundStyle(.secondary)
                if let yearLabel {
                    Text(yearLabel)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }
            }
            .frame(width: 56, alignment: .leading)
            .padding(.top, 4)

            let rows = photos.chunked(into: 3)
            VStack(spacing: 2) {
                ForEach(Array(rows.enumerated()), id: \.offset) { _, rowPhotos in
                    HStack(spacing: 2) {
                        ForEach(rowPhotos) { photo in
                            Button {
                                onImageClick(photo.submissionId)
                            } label: {
                                AsyncImage(url: photo.imageURL) { image in
                                    image.resizable().aspectRatio(contentMode: .fill)
                                } placeholder: {
                                    Color(.secondarySystemGroupedBackground)
                                }
                                .aspectRatio(1, contentMode: .fill)
                                .clipped()
                            }
                            .buttonStyle(.plain)
                            .accessibilityLabel("Photo")
                        }
                        // Fill remaining space in incomplete rows
                        ForEach(0..<(3 - rowPhotos.count), id: \.self) { _ in
                            Color.clear.aspectRatio(1, contentMode: .fill)
                        }
                    }
                }
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
    }
}

extension Array {
    func chunked(into size: Int) -> [[Element]] {
        stride(from: 0, to: count, by: size).map {
            Array(self[$0..<Swift.min($0 + size, count)])
        }
    }
}
